File: go/types/v1.go
package types

import (
	"encoding/json"
)

// PaymentPayloadV1 represents a v1 payment payload structure
// V1 has scheme and network at top level (not in accepted field)
type PaymentPayloadV1 struct {
	X402Version int                    `json:"x402Version"`
	Scheme      string                 `json:"scheme"`
	Network     string                 `json:"network"`
	Payload     map[string]interface{} `json:"payload"`
}

// PaymentPayloadView interface implementation for V1
func (p PaymentPayloadV1) GetVersion() int                    { return p.X402Version }
func (p PaymentPayloadV1) GetScheme() string                  { return p.Scheme }
func (p PaymentPayloadV1) GetNetwork() string                 { return p.Network }
func (p PaymentPayloadV1) GetPayload() map[string]interface{} { return p.Payload }

// PaymentRequirementsV1 represents v1 payment requirements structure
type PaymentRequirementsV1 struct {
	Scheme            string           `json:"scheme"`
	Network           string           `json:"network"`
	MaxAmountRequired string           `json:"maxAmountRequired"`
	Resource          string           `json:"resource"`
	Description       string           `json:"description,omitempty"`
	MimeType          string           `json:"mimeType,omitempty"`
	PayTo             string           `json:"payTo"`
	MaxTimeoutSeconds int              `json:"maxTimeoutSeconds"`
	Asset             string           `json:"asset"`
	OutputSchema      *json.RawMessage `json:"outputSchema,omitempty"`
	Extra             *json.RawMessage `json:"extra,omitempty"`
}

// PaymentRequirementsView interface implementation for V1
func (r PaymentRequirementsV1) GetScheme() string         { return r.Scheme }
func (r PaymentRequirementsV1) GetNetwork() string        { return r.Network }
func (r PaymentRequirementsV1) GetAsset() string          { return r.Asset }
func (r PaymentRequirementsV1) GetAmount() string         { return r.MaxAmountRequired }
func (r PaymentRequirementsV1) GetPayTo() string          { return r.PayTo }
func (r PaymentRequirementsV1) GetMaxTimeoutSeconds() int { return r.MaxTimeoutSeconds }
func (r PaymentRequirementsV1) GetExtra() map[string]interface{} {
	if r.Extra == nil {
		return nil
	}
	var m map[string]interface{}
	if err := json.Unmarshal(*r.Extra, &m); err != nil {
		return make(map[string]interface{}) // Return empty on error
	}
	return m
}

// PaymentRequiredV1 represents a v1 402 response structure
type PaymentRequiredV1 struct {
	X402Version int                     `json:"x402Version"`
	Error       string                  `json:"error,omitempty"`
	Accepts     []PaymentRequirementsV1 `json:"accepts"`
}

// SupportedKindV1 represents a V1 supported payment configuration
type SupportedKindV1 struct {
	X402Version int              `json:"x402Version"`
	Scheme      string           `json:"scheme"`
	Network     string           `json:"network"`
	Extra       *json.RawMessage `json:"extra,omitempty"`
}

// Unmarshal helpers

// ToPaymentPayloadV1 unmarshals bytes to v1 payment payload
func ToPaymentPayloadV1(data []byte) (*PaymentPayloadV1, error) {
	var payload PaymentPayloadV1
	if err := json.Unmarshal(data, &payload); err != nil {
		return nil, err
	}
	return &payload, nil
}

// ToPaymentRequirementsV1 unmarshals bytes to v1 payment requirements
func ToPaymentRequirementsV1(data []byte) (*PaymentRequirementsV1, error) {
	var requirements PaymentRequirementsV1
	if err := json.Unmarshal(data, &requirements); err != nil {
		return nil, err
	}
	return &requirements, nil
}

// ToPaymentRequiredV1 unmarshals bytes to v1 payment required response
func ToPaymentRequiredV1(data []byte) (*PaymentRequiredV1, error) {
	var required PaymentRequiredV1
	if err := json.Unmarshal(data, &required); err != nil {
		return nil, err
	}
	return &required, nil
}

// SupportedResponseV1 is the old supported response format (V1 only, no extensions)
type SupportedResponseV1 struct {
	Kinds []SupportedKindV1 `json:"kinds"`
}

// ToSupportedKindV1 unmarshals bytes to v1 supported kind
func ToSupportedKindV1(data []byte) (*SupportedKindV1, error) {
	var kind SupportedKindV1
	if err := json.Unmarshal(data, &kind); err != nil {
		return nil, err
	}
	return &kind, nil
}

File: python/x402/schemas/v1.py
"""V1 legacy types for the x402 Python SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

from .base import BaseX402Model, Network

if TYPE_CHECKING:
    from .responses import SupportedKind


class PaymentRequirementsV1(BaseX402Model):
    """V1 payment requirements (legacy).

    Attributes:
        scheme: Payment scheme identifier.
        network: Network identifier (legacy format, e.g., "base-sepolia").
        max_amount_required: Maximum amount in smallest unit.
        resource: Resource URL.
        description: Optional resource description.
        mime_type: Optional MIME type.
        pay_to: Recipient address.
        max_timeout_seconds: Maximum time for payment validity.
        asset: Asset address/identifier.
        output_schema: Optional output schema.
        extra: Additional scheme-specific data.
    """

    scheme: str
    network: Network
    max_amount_required: str
    resource: str
    description: str | None = None
    mime_type: str | None = None
    pay_to: str
    max_timeout_seconds: int
    asset: str
    output_schema: dict[str, Any] | None = None
    extra: dict[str, Any] | None = None

    def get_amount(self) -> str:
        """Get the payment amount (V1 uses 'maxAmountRequired' field)."""
        return self.max_amount_required

    def get_extra(self) -> dict[str, Any] | None:
        """Get extra metadata."""
        return self.extra


class PaymentRequiredV1(BaseX402Model):
    """V1 402 response (legacy).

    Attributes:
        x402_version: Protocol version (always 1 for V1).
        error: Optional error message.
        accepts: List of accepted payment requirements.
    """

    x402_version: Literal[1] = 1
    error: str | None = None
    accepts: list[PaymentRequirementsV1]


class PaymentPayloadV1(BaseX402Model):
    """V1 payment payload (legacy).

    Attributes:
        x402_version: Protocol version (always 1 for V1).
        scheme: Payment scheme identifier (at top level in V1).
        network: Network identifier (at top level in V1).
        payload: Scheme-specific payload data.
    """

    x402_version: Literal[1] = 1
    scheme: str
    network: Network
    payload: dict[str, Any]

    def get_scheme(self) -> str:
        """Get the payment scheme (V1 has it at top level)."""
        return self.scheme

    def get_network(self) -> str:
        """Get the network (V1 has it at top level)."""
        return self.network


class SupportedResponseV1(BaseX402Model):
    """V1 supported response (legacy - no extensions or signers).

    Attributes:
        kinds: List of supported payment kinds.
    """

    kinds: list[SupportedKind]
